In [1]:
@file:Repository("*mavenLocal")

In [2]:
%useLatestDescriptors

In [3]:
%use dataframe(v=0.13.0-dev), kandy@d768defdeecace77d118db0f77455970eef4a800

Kotlin DataFrame can be used to work with REST APIs, JSON (let's say some logs or dump of a document database) and other hierarchical structures without the need to write boilerplate code that describes your data.  

Let's see how it works. This example is divided into two sections.
- Section 1. Answer these questions using a prepared data sample:
1. How many contributors there are and how much do they contribute?
2. How many of those contributors are not JetBrains members (according to GitHub)?
3. What is the most starred project?
- Section 2. Build the data sample 

Section 1.

In [4]:
val organizationsRaw = DataFrame.read("https://raw.githubusercontent.com/Kotlin/dataframe/master/data/jetbrains.json")
organizationsRaw

<iframe onload="o_resize_iframe_out_1()" style="width:100%;" class="result_container" id="iframe_out_1" frameBorder="0" srcdoc=" <html theme='dark'>
 <head>
 <style type="text/css">
 :root {
 --background: #fff;
 --background-odd: #f5f5f5;
 --background-hover: #d9edfd;
 --header-text-color: #474747;
 --text-color: #848484;
 --text-color-dark: #000;
 --text-color-medium: #737373;
 --text-color-pale: #b3b3b3;
 --inner-border-color: #aaa;
 --bold-border-color: #000;
 --link-color: #296eaa;
 --link-color-pale: #296eaa;
 --link-hover: #1a466c;
}

:root[theme="dark"], :root [data-jp-theme-light="false"], .dataframe_dark{
 --background: #303030;
 --background-odd: #3c3c3c;
 --background-hover: #464646;
 --header-text-color: #dddddd;
 --text-color: #b3b3b3;
 --text-color-dark: #dddddd;
 --text-color-medium: #b2b2b2;
 --text-color-pale: #737373;
 --inner-border-color: #707070;
 --bold-border-color: #777777;
 --link-color: #008dc0;
 --link-color-pale: #97e1fb;
 --link-hover: #00688e;
}

p.dataframe_description {
 color: var(--text-color-dark);
}

table.dataframe {
 font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
 font-size: 12px;
 background-color: var(--background);
 color: var(--text-color-dark);
 border: none;
 border-collapse: collapse;
}

table.dataframe th, td {
 padding: 6px;
 border: 1px solid transparent;
 text-align: left;
}

table.dataframe th {
 background-color: var(--background);
 color: var(--header-text-color);
}

table.dataframe td {
 vertical-align: top;
}

table.dataframe th.bottomBorder {
 border-bottom-color: var(--bold-border-color);
}

table.dataframe tbody > tr:nth-child(odd) {
 background: var(--background-odd);
}

table.dataframe tbody > tr:nth-child(even) {
 background: var(--background);
}

table.dataframe tbody > tr:hover {
 background: var(--background-hover);
}

table.dataframe a {
 cursor: pointer;
 color: var(--link-color);
 text-decoration: none;
}

table.dataframe tr:hover > td a {
 color: var(--link-color-pale);
}

table.dataframe a:hover {
 color: var(--link-hover);
 text-decoration: underline;
}

table.dataframe img {
 max-width: fit-content;
}

table.dataframe th.complex {
 background-color: var(--background);
 border: 1px solid var(--background);
}

table.dataframe .leftBorder {
 border-left-color: var(--inner-border-color);
}

table.dataframe .rightBorder {
 border-right-color: var(--inner-border-color);
}

table.dataframe .rightAlign {
 text-align: right;
}

table.dataframe .expanderSvg {
 width: 8px;
 height: 8px;
 margin-right: 3px;
}

table.dataframe .expander {
 display: flex;
 align-items: center;
}

/* formatting */

table.dataframe .null {
 color: var(--text-color-pale);
}

table.dataframe .structural {
 color: var(--text-color-medium);
 font-weight: bold;
}

table.dataframe .dataFrameCaption {
 font-weight: bold;
}

table.dataframe .numbers {
 color: var(--text-color-dark);
}

table.dataframe td:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}

table.dataframe tr:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}





:root {
 --scroll-bg: #f5f5f5;
 --scroll-fg: #b3b3b3;
}
:root[theme="dark"], :root [data-jp-theme-light="false"]{
 --scroll-bg: #3c3c3c;
 --scroll-fg: #97e1fb;
}
body {
 scrollbar-color: var(--scroll-fg) var(--scroll-bg);
}
body::-webkit-scrollbar {
 width: 10px; /* Mostly for vertical scrollbars */
 height: 10px; /* Mostly for horizontal scrollbars */
}
body::-webkit-scrollbar-thumb {
 background-color: var(--scroll-fg);
}
body::-webkit-scrollbar-track {
 background-color: var(--scroll-bg);
}
 </style>
 </head>
 <body>
 
<table class="dataframe" id="df_-1426063360"></table>

<table class="dataframe" id="static_df_-1426063352"><thead><tr><th>url</th><th>name</th><th>repos</th><th>members</th></tr></thead><tbody><tr><td>https://api.github.com/orgs/JetBrains</td><td>JetBrains</td><td>[100 x 37]</td><td>[187 x 7]</td></tr></tbody></table>
<p class="dataframe_description">DataFrame: rowsCount = 1, columnsCount = 4</p>

 <

In [5]:
organizationsRaw.schema()

url: String
name: String
repos: *
    id: String
    node_id: String
    name: String
    full_name: String
    private: String
    html_url: String
    description: String?
    fork: String
    url: String
    created_at: String
    updated_at: String
    pushed_at: String
    homepage: String?
    size: String
    stargazers_count: String
    watchers_count: String
    language: String?
    has_issues: String
    has_projects: String
    has_downloads: String
    has_wiki: String
    has_pages: String
    forks_count: String
    archived: String
    disabled: String
    open_issues_count: String
    allow_forking: String
    is_template: String
    topics: String
    visibility: String
    forks: String
    open_issues: String
    watchers: String
    default_branch: String
    permissions:
        admin: String
        maintain: String
        push: String
        triage: String
        pull: String
    contributors: *
        login: String
        id: String
        node_id: String

In [6]:
val organizations = organizationsRaw.parse()
organizations

<iframe onload="o_resize_iframe_out_2()" style="width:100%;" class="result_container" id="iframe_out_2" frameBorder="0" srcdoc=" <html theme='dark'>
 <head>
 <style type="text/css">
 :root {
 --background: #fff;
 --background-odd: #f5f5f5;
 --background-hover: #d9edfd;
 --header-text-color: #474747;
 --text-color: #848484;
 --text-color-dark: #000;
 --text-color-medium: #737373;
 --text-color-pale: #b3b3b3;
 --inner-border-color: #aaa;
 --bold-border-color: #000;
 --link-color: #296eaa;
 --link-color-pale: #296eaa;
 --link-hover: #1a466c;
}

:root[theme="dark"], :root [data-jp-theme-light="false"], .dataframe_dark{
 --background: #303030;
 --background-odd: #3c3c3c;
 --background-hover: #464646;
 --header-text-color: #dddddd;
 --text-color: #b3b3b3;
 --text-color-dark: #dddddd;
 --text-color-medium: #b2b2b2;
 --text-color-pale: #737373;
 --inner-border-color: #707070;
 --bold-border-color: #777777;
 --link-color: #008dc0;
 --link-color-pale: #97e1fb;
 --link-hover: #00688e;
}

p.dataframe_description {
 color: var(--text-color-dark);
}

table.dataframe {
 font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
 font-size: 12px;
 background-color: var(--background);
 color: var(--text-color-dark);
 border: none;
 border-collapse: collapse;
}

table.dataframe th, td {
 padding: 6px;
 border: 1px solid transparent;
 text-align: left;
}

table.dataframe th {
 background-color: var(--background);
 color: var(--header-text-color);
}

table.dataframe td {
 vertical-align: top;
}

table.dataframe th.bottomBorder {
 border-bottom-color: var(--bold-border-color);
}

table.dataframe tbody > tr:nth-child(odd) {
 background: var(--background-odd);
}

table.dataframe tbody > tr:nth-child(even) {
 background: var(--background);
}

table.dataframe tbody > tr:hover {
 background: var(--background-hover);
}

table.dataframe a {
 cursor: pointer;
 color: var(--link-color);
 text-decoration: none;
}

table.dataframe tr:hover > td a {
 color: var(--link-color-pale);
}

table.dataframe a:hover {
 color: var(--link-hover);
 text-decoration: underline;
}

table.dataframe img {
 max-width: fit-content;
}

table.dataframe th.complex {
 background-color: var(--background);
 border: 1px solid var(--background);
}

table.dataframe .leftBorder {
 border-left-color: var(--inner-border-color);
}

table.dataframe .rightBorder {
 border-right-color: var(--inner-border-color);
}

table.dataframe .rightAlign {
 text-align: right;
}

table.dataframe .expanderSvg {
 width: 8px;
 height: 8px;
 margin-right: 3px;
}

table.dataframe .expander {
 display: flex;
 align-items: center;
}

/* formatting */

table.dataframe .null {
 color: var(--text-color-pale);
}

table.dataframe .structural {
 color: var(--text-color-medium);
 font-weight: bold;
}

table.dataframe .dataFrameCaption {
 font-weight: bold;
}

table.dataframe .numbers {
 color: var(--text-color-dark);
}

table.dataframe td:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}

table.dataframe tr:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}





:root {
 --scroll-bg: #f5f5f5;
 --scroll-fg: #b3b3b3;
}
:root[theme="dark"], :root [data-jp-theme-light="false"]{
 --scroll-bg: #3c3c3c;
 --scroll-fg: #97e1fb;
}
body {
 scrollbar-color: var(--scroll-fg) var(--scroll-bg);
}
body::-webkit-scrollbar {
 width: 10px; /* Mostly for vertical scrollbars */
 height: 10px; /* Mostly for horizontal scrollbars */
}
body::-webkit-scrollbar-thumb {
 background-color: var(--scroll-fg);
}
body::-webkit-scrollbar-track {
 background-color: var(--scroll-bg);
}
 </style>
 </head>
 <body>
 
<table class="dataframe" id="df_-1426063351"></table>

<table class="dataframe" id="static_df_-1426063343"><thead><tr><th>url</th><th>name</th><th>repos</th><th>members</th></tr></thead><tbody><tr><td>https://api.github.com/orgs/JetBrains</td><td>JetBrains</td><td>[100 x 37]</td><td>[187 x 7]</td></tr></tbody></table>
<p class="dataframe_description">DataFrame: rowsCount = 1, columnsCount = 4</p>

 <

In [7]:
organizations.schema()

url: URL
name: String
repos: *
    id: Int
    node_id: String
    name: String
    full_name: String
    private: Boolean
    html_url: URL
    description: String?
    fork: Boolean
    url: URL
    created_at: kotlinx.datetime.Instant
    updated_at: kotlinx.datetime.Instant
    pushed_at: kotlinx.datetime.Instant
    homepage: String?
    size: Int
    stargazers_count: Int
    watchers_count: Int
    language: String?
    has_issues: Boolean
    has_projects: Boolean
    has_downloads: Boolean
    has_wiki: Boolean
    has_pages: Boolean
    forks_count: Int
    archived: Boolean
    disabled: Boolean
    open_issues_count: Int
    allow_forking: Boolean
    is_template: Boolean
    topics: String
    visibility: String
    forks: Int
    open_issues: Int
    watchers: Int
    default_branch: String
    permissions:
        admin: Boolean
        maintain: Boolean
        push: Boolean
        triage: Boolean
        pull: Boolean
    contributors: *
        login: String
        

This data is indeed hierarchical. organizations is itself a dataframe. It has a child column `repos`, a dataframe with lots of data.

1. **How many contributors there are?**

The easiest way to work with hierarchical data is to pick only required data from it. It is nice to have a huge dataframe with all relations between data, but carrying it around involves extra complexity. So let's start by getting contributors from inside of the data structure.

In [8]:
val jb = organizations[0]

In [9]:
val jbRepos = jb.repos

In [10]:
jbRepos.contributors

<iframe onload="o_resize_iframe_out_3()" style="width:100%;" class="result_container" id="iframe_out_3" frameBorder="0" srcdoc=" <html theme='dark'>
 <head>
 <style type="text/css">
 :root {
 --background: #fff;
 --background-odd: #f5f5f5;
 --background-hover: #d9edfd;
 --header-text-color: #474747;
 --text-color: #848484;
 --text-color-dark: #000;
 --text-color-medium: #737373;
 --text-color-pale: #b3b3b3;
 --inner-border-color: #aaa;
 --bold-border-color: #000;
 --link-color: #296eaa;
 --link-color-pale: #296eaa;
 --link-hover: #1a466c;
}

:root[theme="dark"], :root [data-jp-theme-light="false"], .dataframe_dark{
 --background: #303030;
 --background-odd: #3c3c3c;
 --background-hover: #464646;
 --header-text-color: #dddddd;
 --text-color: #b3b3b3;
 --text-color-dark: #dddddd;
 --text-color-medium: #b2b2b2;
 --text-color-pale: #737373;
 --inner-border-color: #707070;
 --bold-border-color: #777777;
 --link-color: #008dc0;
 --link-color-pale: #97e1fb;
 --link-hover: #00688e;
}

p.dataframe_description {
 color: var(--text-color-dark);
}

table.dataframe {
 font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
 font-size: 12px;
 background-color: var(--background);
 color: var(--text-color-dark);
 border: none;
 border-collapse: collapse;
}

table.dataframe th, td {
 padding: 6px;
 border: 1px solid transparent;
 text-align: left;
}

table.dataframe th {
 background-color: var(--background);
 color: var(--header-text-color);
}

table.dataframe td {
 vertical-align: top;
}

table.dataframe th.bottomBorder {
 border-bottom-color: var(--bold-border-color);
}

table.dataframe tbody > tr:nth-child(odd) {
 background: var(--background-odd);
}

table.dataframe tbody > tr:nth-child(even) {
 background: var(--background);
}

table.dataframe tbody > tr:hover {
 background: var(--background-hover);
}

table.dataframe a {
 cursor: pointer;
 color: var(--link-color);
 text-decoration: none;
}

table.dataframe tr:hover > td a {
 color: var(--link-color-pale);
}

table.dataframe a:hover {
 color: var(--link-hover);
 text-decoration: underline;
}

table.dataframe img {
 max-width: fit-content;
}

table.dataframe th.complex {
 background-color: var(--background);
 border: 1px solid var(--background);
}

table.dataframe .leftBorder {
 border-left-color: var(--inner-border-color);
}

table.dataframe .rightBorder {
 border-right-color: var(--inner-border-color);
}

table.dataframe .rightAlign {
 text-align: right;
}

table.dataframe .expanderSvg {
 width: 8px;
 height: 8px;
 margin-right: 3px;
}

table.dataframe .expander {
 display: flex;
 align-items: center;
}

/* formatting */

table.dataframe .null {
 color: var(--text-color-pale);
}

table.dataframe .structural {
 color: var(--text-color-medium);
 font-weight: bold;
}

table.dataframe .dataFrameCaption {
 font-weight: bold;
}

table.dataframe .numbers {
 color: var(--text-color-dark);
}

table.dataframe td:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}

table.dataframe tr:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}





:root {
 --scroll-bg: #f5f5f5;
 --scroll-fg: #b3b3b3;
}
:root[theme="dark"], :root [data-jp-theme-light="false"]{
 --scroll-bg: #3c3c3c;
 --scroll-fg: #97e1fb;
}
body {
 scrollbar-color: var(--scroll-fg) var(--scroll-bg);
}
body::-webkit-scrollbar {
 width: 10px; /* Mostly for vertical scrollbars */
 height: 10px; /* Mostly for horizontal scrollbars */
}
body::-webkit-scrollbar-thumb {
 background-color: var(--scroll-fg);
}
body::-webkit-scrollbar-track {
 background-color: var(--scroll-bg);
}
 </style>
 </head>
 <body>
 
<table class="dataframe" id="df_-1426063342"></table>

<table class="dataframe" id="static_df_-1426063321"><thead><tr><th>contributors</th></tr></thead><tbody><tr><td>[4 x 8]</td></tr><tr><td>[20 x 8]</td></tr><tr><td>[10 x 8]</td></tr><tr><td>[87 x 8]</td></tr><tr><td>[1 x 8] { login:anna239, id:661533, n<span class="structural">...</span></td></tr><tr><td>[1 x 8] { login:anna239, id:661533, n<span 

So `contributors` is a [FrameColumn](https://kotlin.github.io/dataframe/datacolumn.html#framecolumn), a column of dataframes. There is a family of operations [to work with several dataframes](https://kotlin.github.io/dataframe/multipledataframes.html). We can use [`concat`](https://kotlin.github.io/dataframe/concat.html) to simplify the structure.

In [11]:
val jbContributors = jbRepos.contributors.concat()
jbContributors

<iframe onload="o_resize_iframe_out_4()" style="width:100%;" class="result_container" id="iframe_out_4" frameBorder="0" srcdoc=" <html theme='dark'>
 <head>
 <style type="text/css">
 :root {
 --background: #fff;
 --background-odd: #f5f5f5;
 --background-hover: #d9edfd;
 --header-text-color: #474747;
 --text-color: #848484;
 --text-color-dark: #000;
 --text-color-medium: #737373;
 --text-color-pale: #b3b3b3;
 --inner-border-color: #aaa;
 --bold-border-color: #000;
 --link-color: #296eaa;
 --link-color-pale: #296eaa;
 --link-hover: #1a466c;
}

:root[theme="dark"], :root [data-jp-theme-light="false"], .dataframe_dark{
 --background: #303030;
 --background-odd: #3c3c3c;
 --background-hover: #464646;
 --header-text-color: #dddddd;
 --text-color: #b3b3b3;
 --text-color-dark: #dddddd;
 --text-color-medium: #b2b2b2;
 --text-color-pale: #737373;
 --inner-border-color: #707070;
 --bold-border-color: #777777;
 --link-color: #008dc0;
 --link-color-pale: #97e1fb;
 --link-hover: #00688e;
}

p.dataframe_description {
 color: var(--text-color-dark);
}

table.dataframe {
 font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
 font-size: 12px;
 background-color: var(--background);
 color: var(--text-color-dark);
 border: none;
 border-collapse: collapse;
}

table.dataframe th, td {
 padding: 6px;
 border: 1px solid transparent;
 text-align: left;
}

table.dataframe th {
 background-color: var(--background);
 color: var(--header-text-color);
}

table.dataframe td {
 vertical-align: top;
}

table.dataframe th.bottomBorder {
 border-bottom-color: var(--bold-border-color);
}

table.dataframe tbody > tr:nth-child(odd) {
 background: var(--background-odd);
}

table.dataframe tbody > tr:nth-child(even) {
 background: var(--background);
}

table.dataframe tbody > tr:hover {
 background: var(--background-hover);
}

table.dataframe a {
 cursor: pointer;
 color: var(--link-color);
 text-decoration: none;
}

table.dataframe tr:hover > td a {
 color: var(--link-color-pale);
}

table.dataframe a:hover {
 color: var(--link-hover);
 text-decoration: underline;
}

table.dataframe img {
 max-width: fit-content;
}

table.dataframe th.complex {
 background-color: var(--background);
 border: 1px solid var(--background);
}

table.dataframe .leftBorder {
 border-left-color: var(--inner-border-color);
}

table.dataframe .rightBorder {
 border-right-color: var(--inner-border-color);
}

table.dataframe .rightAlign {
 text-align: right;
}

table.dataframe .expanderSvg {
 width: 8px;
 height: 8px;
 margin-right: 3px;
}

table.dataframe .expander {
 display: flex;
 align-items: center;
}

/* formatting */

table.dataframe .null {
 color: var(--text-color-pale);
}

table.dataframe .structural {
 color: var(--text-color-medium);
 font-weight: bold;
}

table.dataframe .dataFrameCaption {
 font-weight: bold;
}

table.dataframe .numbers {
 color: var(--text-color-dark);
}

table.dataframe td:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}

table.dataframe tr:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}





:root {
 --scroll-bg: #f5f5f5;
 --scroll-fg: #b3b3b3;
}
:root[theme="dark"], :root [data-jp-theme-light="false"]{
 --scroll-bg: #3c3c3c;
 --scroll-fg: #97e1fb;
}
body {
 scrollbar-color: var(--scroll-fg) var(--scroll-bg);
}
body::-webkit-scrollbar {
 width: 10px; /* Mostly for vertical scrollbars */
 height: 10px; /* Mostly for horizontal scrollbars */
}
body::-webkit-scrollbar-thumb {
 background-color: var(--scroll-fg);
}
body::-webkit-scrollbar-track {
 background-color: var(--scroll-bg);
}
 </style>
 </head>
 <body>
 
<table class="dataframe" id="df_-1426063320"></table>

<table class="dataframe" id="static_df_-1426063319"><thead><tr><th>login</th><th>id</th><th>node_id</th><th>gravatar_id</th><th>url</th><th>type</th><th>site_admin</th><th>contributions</th></tr></thead><tbody><tr><td>chashnikov</td><td>978014</td><td>MDQ6VXNlcjk3ODAxNA==</td><td></td><td>https://api.github.com/users/chashnikov</td><td>User</td><

Probably not all of them are unique rows. Let's use [`distinctBy`](https://kotlin.github.io/dataframe/distinct.html#distinctby)

In [12]:
jbContributors.distinctBy { login }.count()

1123

So, the answer to the question "How many contributors there are" is 1123.

Now that we have a dataframe with all contributors from jb repositories, let's see **how much do they contribute** with the help of [groupBy aggregations](https://kotlin.github.io/dataframe/groupby.html#aggregation)

In [13]:
val top = jbContributors.groupBy { login and url }.sum { contributions }.sortByDesc { contributions }

2. How many of those contributors are not JetBrains members (according to GitHub)?

In [14]:
val memberIds = jb.members.id.toSet()
memberIds.size

187

In [15]:
val communityContributors = jbContributors.filter { it.id !in memberIds }

In [16]:
communityContributors.groupBy { login and url }.sum { contributions }.sortByDesc { contributions }

<iframe onload="o_resize_iframe_out_5()" style="width:100%;" class="result_container" id="iframe_out_5" frameBorder="0" srcdoc=" <html theme='dark'>
 <head>
 <style type="text/css">
 :root {
 --background: #fff;
 --background-odd: #f5f5f5;
 --background-hover: #d9edfd;
 --header-text-color: #474747;
 --text-color: #848484;
 --text-color-dark: #000;
 --text-color-medium: #737373;
 --text-color-pale: #b3b3b3;
 --inner-border-color: #aaa;
 --bold-border-color: #000;
 --link-color: #296eaa;
 --link-color-pale: #296eaa;
 --link-hover: #1a466c;
}

:root[theme="dark"], :root [data-jp-theme-light="false"], .dataframe_dark{
 --background: #303030;
 --background-odd: #3c3c3c;
 --background-hover: #464646;
 --header-text-color: #dddddd;
 --text-color: #b3b3b3;
 --text-color-dark: #dddddd;
 --text-color-medium: #b2b2b2;
 --text-color-pale: #737373;
 --inner-border-color: #707070;
 --bold-border-color: #777777;
 --link-color: #008dc0;
 --link-color-pale: #97e1fb;
 --link-hover: #00688e;
}

p.dataframe_description {
 color: var(--text-color-dark);
}

table.dataframe {
 font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
 font-size: 12px;
 background-color: var(--background);
 color: var(--text-color-dark);
 border: none;
 border-collapse: collapse;
}

table.dataframe th, td {
 padding: 6px;
 border: 1px solid transparent;
 text-align: left;
}

table.dataframe th {
 background-color: var(--background);
 color: var(--header-text-color);
}

table.dataframe td {
 vertical-align: top;
}

table.dataframe th.bottomBorder {
 border-bottom-color: var(--bold-border-color);
}

table.dataframe tbody > tr:nth-child(odd) {
 background: var(--background-odd);
}

table.dataframe tbody > tr:nth-child(even) {
 background: var(--background);
}

table.dataframe tbody > tr:hover {
 background: var(--background-hover);
}

table.dataframe a {
 cursor: pointer;
 color: var(--link-color);
 text-decoration: none;
}

table.dataframe tr:hover > td a {
 color: var(--link-color-pale);
}

table.dataframe a:hover {
 color: var(--link-hover);
 text-decoration: underline;
}

table.dataframe img {
 max-width: fit-content;
}

table.dataframe th.complex {
 background-color: var(--background);
 border: 1px solid var(--background);
}

table.dataframe .leftBorder {
 border-left-color: var(--inner-border-color);
}

table.dataframe .rightBorder {
 border-right-color: var(--inner-border-color);
}

table.dataframe .rightAlign {
 text-align: right;
}

table.dataframe .expanderSvg {
 width: 8px;
 height: 8px;
 margin-right: 3px;
}

table.dataframe .expander {
 display: flex;
 align-items: center;
}

/* formatting */

table.dataframe .null {
 color: var(--text-color-pale);
}

table.dataframe .structural {
 color: var(--text-color-medium);
 font-weight: bold;
}

table.dataframe .dataFrameCaption {
 font-weight: bold;
}

table.dataframe .numbers {
 color: var(--text-color-dark);
}

table.dataframe td:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}

table.dataframe tr:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}





:root {
 --scroll-bg: #f5f5f5;
 --scroll-fg: #b3b3b3;
}
:root[theme="dark"], :root [data-jp-theme-light="false"]{
 --scroll-bg: #3c3c3c;
 --scroll-fg: #97e1fb;
}
body {
 scrollbar-color: var(--scroll-fg) var(--scroll-bg);
}
body::-webkit-scrollbar {
 width: 10px; /* Mostly for vertical scrollbars */
 height: 10px; /* Mostly for horizontal scrollbars */
}
body::-webkit-scrollbar-thumb {
 background-color: var(--scroll-fg);
}
body::-webkit-scrollbar-track {
 background-color: var(--scroll-bg);
}
 </style>
 </head>
 <body>
 
<table class="dataframe" id="df_-1426063318"></table>

<table class="dataframe" id="static_df_-1426063317"><thead><tr><th>login</th><th>url</th><th>contributions</th></tr></thead><tbody><tr><td>akozlova</td><td>https://api.github.com/users/akozlova</td><td>21886</td></tr><tr><td>cdracm</td><td>https://api.github.com/users/cdracm</td><td>15667</td></tr><tr><td>develar</td><td>https://api.github.com/us

Let's visualise the difference two groups of contributors: members and non-members. We will explain `map` operation in more details later

In [17]:
val df = jbRepos.mapToFrame { 
    "id" from { id }
    +name
    "total" from  { contributors.count() }
    "community" from { contributors.filter { it.id !in memberIds }.count() }
}
df

<iframe onload="o_resize_iframe_out_6()" style="width:100%;" class="result_container" id="iframe_out_6" frameBorder="0" srcdoc=" <html theme='dark'>
 <head>
 <style type="text/css">
 :root {
 --background: #fff;
 --background-odd: #f5f5f5;
 --background-hover: #d9edfd;
 --header-text-color: #474747;
 --text-color: #848484;
 --text-color-dark: #000;
 --text-color-medium: #737373;
 --text-color-pale: #b3b3b3;
 --inner-border-color: #aaa;
 --bold-border-color: #000;
 --link-color: #296eaa;
 --link-color-pale: #296eaa;
 --link-hover: #1a466c;
}

:root[theme="dark"], :root [data-jp-theme-light="false"], .dataframe_dark{
 --background: #303030;
 --background-odd: #3c3c3c;
 --background-hover: #464646;
 --header-text-color: #dddddd;
 --text-color: #b3b3b3;
 --text-color-dark: #dddddd;
 --text-color-medium: #b2b2b2;
 --text-color-pale: #737373;
 --inner-border-color: #707070;
 --bold-border-color: #777777;
 --link-color: #008dc0;
 --link-color-pale: #97e1fb;
 --link-hover: #00688e;
}

p.dataframe_description {
 color: var(--text-color-dark);
}

table.dataframe {
 font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
 font-size: 12px;
 background-color: var(--background);
 color: var(--text-color-dark);
 border: none;
 border-collapse: collapse;
}

table.dataframe th, td {
 padding: 6px;
 border: 1px solid transparent;
 text-align: left;
}

table.dataframe th {
 background-color: var(--background);
 color: var(--header-text-color);
}

table.dataframe td {
 vertical-align: top;
}

table.dataframe th.bottomBorder {
 border-bottom-color: var(--bold-border-color);
}

table.dataframe tbody > tr:nth-child(odd) {
 background: var(--background-odd);
}

table.dataframe tbody > tr:nth-child(even) {
 background: var(--background);
}

table.dataframe tbody > tr:hover {
 background: var(--background-hover);
}

table.dataframe a {
 cursor: pointer;
 color: var(--link-color);
 text-decoration: none;
}

table.dataframe tr:hover > td a {
 color: var(--link-color-pale);
}

table.dataframe a:hover {
 color: var(--link-hover);
 text-decoration: underline;
}

table.dataframe img {
 max-width: fit-content;
}

table.dataframe th.complex {
 background-color: var(--background);
 border: 1px solid var(--background);
}

table.dataframe .leftBorder {
 border-left-color: var(--inner-border-color);
}

table.dataframe .rightBorder {
 border-right-color: var(--inner-border-color);
}

table.dataframe .rightAlign {
 text-align: right;
}

table.dataframe .expanderSvg {
 width: 8px;
 height: 8px;
 margin-right: 3px;
}

table.dataframe .expander {
 display: flex;
 align-items: center;
}

/* formatting */

table.dataframe .null {
 color: var(--text-color-pale);
}

table.dataframe .structural {
 color: var(--text-color-medium);
 font-weight: bold;
}

table.dataframe .dataFrameCaption {
 font-weight: bold;
}

table.dataframe .numbers {
 color: var(--text-color-dark);
}

table.dataframe td:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}

table.dataframe tr:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}





:root {
 --scroll-bg: #f5f5f5;
 --scroll-fg: #b3b3b3;
}
:root[theme="dark"], :root [data-jp-theme-light="false"]{
 --scroll-bg: #3c3c3c;
 --scroll-fg: #97e1fb;
}
body {
 scrollbar-color: var(--scroll-fg) var(--scroll-bg);
}
body::-webkit-scrollbar {
 width: 10px; /* Mostly for vertical scrollbars */
 height: 10px; /* Mostly for horizontal scrollbars */
}
body::-webkit-scrollbar-thumb {
 background-color: var(--scroll-fg);
}
body::-webkit-scrollbar-track {
 background-color: var(--scroll-bg);
}
 </style>
 </head>
 <body>
 
<table class="dataframe" id="df_-1426063316"></table>

<table class="dataframe" id="static_df_-1426063315"><thead><tr><th>id</th><th>name</th><th>total</th><th>community</th></tr></thead><tbody><tr><td>294457</td><td>JPS</td><td>4</td><td>0</td></tr><tr><td>1098358</td><td>YouTrackSharp</td><td>20</td><td>17</td></tr><tr><td>1156792</td><td>colorSchemeTool</td><td>10</td><td>7</td></tr><tr><td>145948

In [18]:
df.plot { 
    x(name)
    step { 
        y(total) { axis.name = "contrib "}
    }
    step { 
        y(community)
        color = Color.named("dark_green")
    }
    layout.size = 1200 to 500
}

JPS
 
 
 
 
 
 
 
 
 
 
 
 
 colorSchemeTool
 
 
 
 
 
 
 
 
 
 
 
 
 youtrack-vcs-hooks
 
 
 
 
 
 
 
 
 
 
 
 
 emacs4ij
 
 
 
 
 
 
 
 
 
 
 
 
 teamcity-nuget-support
 
 
 
 
 
 
 
 
 
 
 
 
 intellij-starteam-plugin
 
 
 
 
 
 
 
 
 
 
 
 
 MPS
 
 
 
 
 
 
 
 
 
 
 
 
 TeamCity.ServiceMessages
 
 
 
 
 
 
 
 
 
 
 
 
 intellij-scala
 
 
 
 
 
 
 
 
 
 
 
 
 teamcity-cpp
 
 
 
 
 
 
 
 
 
 
 
 
 teamcity-local-cloud
 
 
 
 
 
 
 
 
 
 
 
 
 intellij-plugin-verifier
 
 
 
 
 
 
 
 
 
 
 
 
 dottrace-demos
 
 
 
 
 
 
 
 
 
 
 
 
 intellij-plugins
 
 
 
 
 
 
 
 
 
 
 
 
 teamcity-shared-resources
 
 
 
 
 
 
 
 
 
 
 
 
 kannotator
 
 
 
 
 
 
 
 
 
 
 
 
 mnemonics
 
 
 
 
 
 
 
 
 
 
 
 
 intellij-coffee-script-debugger
 
 
 
 
 
 
 
 
 
 
 
 
 kara
 
 
 
 
 
 
 
 
 
 
 
 
 hsql1
 
 
 
 
 
 
 
 
 
 
 
 
 resharper-angularjs
 
 
 
 
 
 
 
 
 
 
 
 
 kotlin-eclipse
 
 
 
 
 
 
 
 
 
 
 
 
 ReSharperGallery
 
 
 
 
 
 
 
 
 
 
 
 
 ReTrack
 
 
 
 
 
 
 
 
 
 
 
 
 dotpeek-nupeek
 
 
 
 
 
 
 
 
 
 
 
 
 JetBox
 
 
 
 
 
 
 
 
 
 
 
 
 intellij-feature-suggester
 
 
 
 
 
 
 
 
 
 
 
 
 dotnet-cas-client
 
 
 
 
 
 
 
 
 
 
 
 
 intellij-haxe
 
 
 
 
 
 
 
 
 
 
 
 
 dekaf
 
 
 
 
 
 
 
 
 
 
 
 
 pty4j
 
 
 
 
 
 
 
 
 
 
 
 
 Exposed
 
 
 
 
 
 
 
 
 
 
 
 
 NuGetOperations
 
 
 
 
 
 
 
 
 
 
 
 
 youtrack.mylyn.connector
 
 
 
 
 
 
 
 
 
 
 
 
 dotpeek-peekfx
 
 
 
 
 
 
 
 
 
 
 
 
 TeamCity.StaticUIExtensions
 
 
 
 
 
 
 
 
 
 
 
 
 sbt-structure
 
 
 
 
 
 
 
 
 
 
 
 
 resharper-phantomjs
 
 
 
 
 
 
 
 
 
 
 
 
 intellij-tutorials-december-2013
 
 
 
 
 
 
 
 
 
 
 
 
 tc-sbt-runner
 
 
 
 
 
 
 
 
 
 
 
 
 phpstorm-workshop
 
 
 
 
 
 
 
 
 
 
 
 
 integration-tests-vm-scripts
 
 
 
 
 
 
 
 
 
 
 
 
 TeamCity.PluginTemplate
 
 
 
 
 
 
 
 
 
 
 
 
 dotpeek-console
 
 
 
 
 
 
 
 
 
 
 
 
 teamcity-sdk-maven-plugin
 
 
 
 
 
 
 
 
 
 
 
 
 teamcity-achievements
 
 
 
 
 
 
 
 
 
 
 
 
 intellij-ipnb
 
 
 
 
 
 
 
 
 
 
 
 
 TeamCity.Widgets
 
 
 
 
 
 
 
 
 
 
 
 
 resharper-devguide
 
 
 
 
 
 
 
 
 
 
 
 
 teamcity-tests-watchdog
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 0
 
 
 
 
 
 
 50
 
 
 
 
 
 
 100
 
 
 
 
 
 
 
 
 
 
 
 
 <path d="M2.287843427834196 174.74009149144848 L2.287843427834196 174.74009149144848 L13.727060567005175 174.74009149144848 L13.727060567005175 147.05849283933782 L25.166277706176157 147.05849283933782 L25.166277706176157 164.359491996907 L36.60549484534714 164.359491996907 L36.60549484534714 31.141798483624484 L48.044711984518116 31.141798483624484 L48.044711984518116 179.93039123871924 L59.483929123689094 179.93039123871924 L59.483929123689094 179.93039123871924 L70.92314626286007 179.93039123871924 L70.92314626286007 176.4701914072054 L82.36236340203105 176.4701914072054 L82.36236340203105 179.93039123871924 L93.80158054120203 179.93039123871924 L93.80158054120203 145.3283929235809 L105.240797680373 145.3283929235809 L105.240797680373 160.89929216539315 L116.68001481954398 160.89929216539315 L116.68001481954398 179.93039123871924 L128.11923195871498 179.93039123871924 L128.11923195871498 134.94779342903942 L139.55844909788595 134.94779342903942 L139.55844909788595 107.26619477692877 L150.99766623705693 107.26619477692877 L150.99766623705693 8.650499578784576 L162.4368833762279 8.650499578784576 L162.4368833762279 164.359491996907 L173.8761005153989 164.359491996907 L173.8761005153989 136.67789334479633 L185.31531765456987 136.67789334479633 L185.31531765456987 8.650499578784576 L196.75453479374085 8.650499578784576 L196.75453479374085 133.2176935132825 L208.19375193291182 133.2176935132825 L208.19375193291182 167.81969182842082 L219.6329690720828 167.81969182842082 L219.6329690720828 8.650499578784576 L231.07218621125378 8.650499578784576 L231.07218621125378 176.4701914072054 L242.51140335042476 176.4701914072054 L242.51140335042476 131.4875935975256 L253.95062048959574 131.4875935975256 L253.95062048959574 148.78859275509475 L265.38983762876677 148.78859275509475 L265.38983762876677 

3. **What is the most starred project?** Naturally, as our dataframe has only one row, we can find it out like this

In [19]:
jbRepos.maxBy { stargazers_count }

<iframe onload="o_resize_iframe_out_7()" style="width:100%;" class="result_container" id="iframe_out_7" frameBorder="0" srcdoc=" <html theme='dark'>
 <head>
 <style type="text/css">
 :root {
 --background: #fff;
 --background-odd: #f5f5f5;
 --background-hover: #d9edfd;
 --header-text-color: #474747;
 --text-color: #848484;
 --text-color-dark: #000;
 --text-color-medium: #737373;
 --text-color-pale: #b3b3b3;
 --inner-border-color: #aaa;
 --bold-border-color: #000;
 --link-color: #296eaa;
 --link-color-pale: #296eaa;
 --link-hover: #1a466c;
}

:root[theme="dark"], :root [data-jp-theme-light="false"], .dataframe_dark{
 --background: #303030;
 --background-odd: #3c3c3c;
 --background-hover: #464646;
 --header-text-color: #dddddd;
 --text-color: #b3b3b3;
 --text-color-dark: #dddddd;
 --text-color-medium: #b2b2b2;
 --text-color-pale: #737373;
 --inner-border-color: #707070;
 --bold-border-color: #777777;
 --link-color: #008dc0;
 --link-color-pale: #97e1fb;
 --link-hover: #00688e;
}

p.dataframe_description {
 color: var(--text-color-dark);
}

table.dataframe {
 font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
 font-size: 12px;
 background-color: var(--background);
 color: var(--text-color-dark);
 border: none;
 border-collapse: collapse;
}

table.dataframe th, td {
 padding: 6px;
 border: 1px solid transparent;
 text-align: left;
}

table.dataframe th {
 background-color: var(--background);
 color: var(--header-text-color);
}

table.dataframe td {
 vertical-align: top;
}

table.dataframe th.bottomBorder {
 border-bottom-color: var(--bold-border-color);
}

table.dataframe tbody > tr:nth-child(odd) {
 background: var(--background-odd);
}

table.dataframe tbody > tr:nth-child(even) {
 background: var(--background);
}

table.dataframe tbody > tr:hover {
 background: var(--background-hover);
}

table.dataframe a {
 cursor: pointer;
 color: var(--link-color);
 text-decoration: none;
}

table.dataframe tr:hover > td a {
 color: var(--link-color-pale);
}

table.dataframe a:hover {
 color: var(--link-hover);
 text-decoration: underline;
}

table.dataframe img {
 max-width: fit-content;
}

table.dataframe th.complex {
 background-color: var(--background);
 border: 1px solid var(--background);
}

table.dataframe .leftBorder {
 border-left-color: var(--inner-border-color);
}

table.dataframe .rightBorder {
 border-right-color: var(--inner-border-color);
}

table.dataframe .rightAlign {
 text-align: right;
}

table.dataframe .expanderSvg {
 width: 8px;
 height: 8px;
 margin-right: 3px;
}

table.dataframe .expander {
 display: flex;
 align-items: center;
}

/* formatting */

table.dataframe .null {
 color: var(--text-color-pale);
}

table.dataframe .structural {
 color: var(--text-color-medium);
 font-weight: bold;
}

table.dataframe .dataFrameCaption {
 font-weight: bold;
}

table.dataframe .numbers {
 color: var(--text-color-dark);
}

table.dataframe td:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}

table.dataframe tr:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}





:root {
 --scroll-bg: #f5f5f5;
 --scroll-fg: #b3b3b3;
}
:root[theme="dark"], :root [data-jp-theme-light="false"]{
 --scroll-bg: #3c3c3c;
 --scroll-fg: #97e1fb;
}
body {
 scrollbar-color: var(--scroll-fg) var(--scroll-bg);
}
body::-webkit-scrollbar {
 width: 10px; /* Mostly for vertical scrollbars */
 height: 10px; /* Mostly for horizontal scrollbars */
}
body::-webkit-scrollbar-thumb {
 background-color: var(--scroll-fg);
}
body::-webkit-scrollbar-track {
 background-color: var(--scroll-bg);
}
 </style>
 </head>
 <body>
 
<table class="dataframe" id="df_-1426063314"></table>

<table class="dataframe" id="static_df_-1426063312"><thead><tr><th>id</th><th>node_id</th><th>name</th><th>full_name</th><th>private</th><th>html_url</th><th>description</th><th>fork</th><th>url</th><th>created_at</th><th>updated_at</th><th>pushed_at</th><th>homepage</th><th>size</th><th>stargazers_count</th><th>watchers_count</th><th>language</t

But usually dataframe has several rows. With dataframe, you can work with hierarchical data structures as a whole, if you need to. Let's use `map` to find a repository with the most stars for each row in the `jb` dataframe. The result of `map` is a data column `most_starred_repo`. 

In [20]:
organizations.mapToColumn("most_starred_repo") {
    it.repos.maxBy { stargazers_count }
}

<iframe onload="o_resize_iframe_out_8()" style="width:100%;" class="result_container" id="iframe_out_8" frameBorder="0" srcdoc=" <html theme='dark'>
 <head>
 <style type="text/css">
 :root {
 --background: #fff;
 --background-odd: #f5f5f5;
 --background-hover: #d9edfd;
 --header-text-color: #474747;
 --text-color: #848484;
 --text-color-dark: #000;
 --text-color-medium: #737373;
 --text-color-pale: #b3b3b3;
 --inner-border-color: #aaa;
 --bold-border-color: #000;
 --link-color: #296eaa;
 --link-color-pale: #296eaa;
 --link-hover: #1a466c;
}

:root[theme="dark"], :root [data-jp-theme-light="false"], .dataframe_dark{
 --background: #303030;
 --background-odd: #3c3c3c;
 --background-hover: #464646;
 --header-text-color: #dddddd;
 --text-color: #b3b3b3;
 --text-color-dark: #dddddd;
 --text-color-medium: #b2b2b2;
 --text-color-pale: #737373;
 --inner-border-color: #707070;
 --bold-border-color: #777777;
 --link-color: #008dc0;
 --link-color-pale: #97e1fb;
 --link-hover: #00688e;
}

p.dataframe_description {
 color: var(--text-color-dark);
}

table.dataframe {
 font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
 font-size: 12px;
 background-color: var(--background);
 color: var(--text-color-dark);
 border: none;
 border-collapse: collapse;
}

table.dataframe th, td {
 padding: 6px;
 border: 1px solid transparent;
 text-align: left;
}

table.dataframe th {
 background-color: var(--background);
 color: var(--header-text-color);
}

table.dataframe td {
 vertical-align: top;
}

table.dataframe th.bottomBorder {
 border-bottom-color: var(--bold-border-color);
}

table.dataframe tbody > tr:nth-child(odd) {
 background: var(--background-odd);
}

table.dataframe tbody > tr:nth-child(even) {
 background: var(--background);
}

table.dataframe tbody > tr:hover {
 background: var(--background-hover);
}

table.dataframe a {
 cursor: pointer;
 color: var(--link-color);
 text-decoration: none;
}

table.dataframe tr:hover > td a {
 color: var(--link-color-pale);
}

table.dataframe a:hover {
 color: var(--link-hover);
 text-decoration: underline;
}

table.dataframe img {
 max-width: fit-content;
}

table.dataframe th.complex {
 background-color: var(--background);
 border: 1px solid var(--background);
}

table.dataframe .leftBorder {
 border-left-color: var(--inner-border-color);
}

table.dataframe .rightBorder {
 border-right-color: var(--inner-border-color);
}

table.dataframe .rightAlign {
 text-align: right;
}

table.dataframe .expanderSvg {
 width: 8px;
 height: 8px;
 margin-right: 3px;
}

table.dataframe .expander {
 display: flex;
 align-items: center;
}

/* formatting */

table.dataframe .null {
 color: var(--text-color-pale);
}

table.dataframe .structural {
 color: var(--text-color-medium);
 font-weight: bold;
}

table.dataframe .dataFrameCaption {
 font-weight: bold;
}

table.dataframe .numbers {
 color: var(--text-color-dark);
}

table.dataframe td:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}

table.dataframe tr:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}





:root {
 --scroll-bg: #f5f5f5;
 --scroll-fg: #b3b3b3;
}
:root[theme="dark"], :root [data-jp-theme-light="false"]{
 --scroll-bg: #3c3c3c;
 --scroll-fg: #97e1fb;
}
body {
 scrollbar-color: var(--scroll-fg) var(--scroll-bg);
}
body::-webkit-scrollbar {
 width: 10px; /* Mostly for vertical scrollbars */
 height: 10px; /* Mostly for horizontal scrollbars */
}
body::-webkit-scrollbar-thumb {
 background-color: var(--scroll-fg);
}
body::-webkit-scrollbar-track {
 background-color: var(--scroll-bg);
}
 </style>
 </head>
 <body>
 
<table class="dataframe" id="df_-1426063311"></table>

<table class="dataframe" id="static_df_-1426063309"><thead><tr><th>most_starred_repo</th></tr></thead><tbody><tr><td>{ id:3432266, node_id:MDEwOlJlcG9zaXR<span class="structural">...</span></td></tr></tbody></table>
<p class="dataframe_description">DataFrame: rowsCount = 1, columnsCount = 37</p>

 </body>
 <script>
 (function () {
 window.DataFra

Expressions inside map, a row expression (docs), enable two different syntax to refer to data. **You can write `this.repos`, `it.repos` or just `repos`. It's all the same, but let's use shorter syntax from now on.**

In [21]:
organizations.mapToColumn("most_starred_repo") {
    // Alternative syntax
    // this.repos.maxBy { stargazers_count }
    // it.repos.maxBy { stargazers_count }
    repos.maxBy { stargazers_count }
}

<iframe onload="o_resize_iframe_out_9()" style="width:100%;" class="result_container" id="iframe_out_9" frameBorder="0" srcdoc=" <html theme='dark'>
 <head>
 <style type="text/css">
 :root {
 --background: #fff;
 --background-odd: #f5f5f5;
 --background-hover: #d9edfd;
 --header-text-color: #474747;
 --text-color: #848484;
 --text-color-dark: #000;
 --text-color-medium: #737373;
 --text-color-pale: #b3b3b3;
 --inner-border-color: #aaa;
 --bold-border-color: #000;
 --link-color: #296eaa;
 --link-color-pale: #296eaa;
 --link-hover: #1a466c;
}

:root[theme="dark"], :root [data-jp-theme-light="false"], .dataframe_dark{
 --background: #303030;
 --background-odd: #3c3c3c;
 --background-hover: #464646;
 --header-text-color: #dddddd;
 --text-color: #b3b3b3;
 --text-color-dark: #dddddd;
 --text-color-medium: #b2b2b2;
 --text-color-pale: #737373;
 --inner-border-color: #707070;
 --bold-border-color: #777777;
 --link-color: #008dc0;
 --link-color-pale: #97e1fb;
 --link-hover: #00688e;
}

p.dataframe_description {
 color: var(--text-color-dark);
}

table.dataframe {
 font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
 font-size: 12px;
 background-color: var(--background);
 color: var(--text-color-dark);
 border: none;
 border-collapse: collapse;
}

table.dataframe th, td {
 padding: 6px;
 border: 1px solid transparent;
 text-align: left;
}

table.dataframe th {
 background-color: var(--background);
 color: var(--header-text-color);
}

table.dataframe td {
 vertical-align: top;
}

table.dataframe th.bottomBorder {
 border-bottom-color: var(--bold-border-color);
}

table.dataframe tbody > tr:nth-child(odd) {
 background: var(--background-odd);
}

table.dataframe tbody > tr:nth-child(even) {
 background: var(--background);
}

table.dataframe tbody > tr:hover {
 background: var(--background-hover);
}

table.dataframe a {
 cursor: pointer;
 color: var(--link-color);
 text-decoration: none;
}

table.dataframe tr:hover > td a {
 color: var(--link-color-pale);
}

table.dataframe a:hover {
 color: var(--link-hover);
 text-decoration: underline;
}

table.dataframe img {
 max-width: fit-content;
}

table.dataframe th.complex {
 background-color: var(--background);
 border: 1px solid var(--background);
}

table.dataframe .leftBorder {
 border-left-color: var(--inner-border-color);
}

table.dataframe .rightBorder {
 border-right-color: var(--inner-border-color);
}

table.dataframe .rightAlign {
 text-align: right;
}

table.dataframe .expanderSvg {
 width: 8px;
 height: 8px;
 margin-right: 3px;
}

table.dataframe .expander {
 display: flex;
 align-items: center;
}

/* formatting */

table.dataframe .null {
 color: var(--text-color-pale);
}

table.dataframe .structural {
 color: var(--text-color-medium);
 font-weight: bold;
}

table.dataframe .dataFrameCaption {
 font-weight: bold;
}

table.dataframe .numbers {
 color: var(--text-color-dark);
}

table.dataframe td:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}

table.dataframe tr:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}





:root {
 --scroll-bg: #f5f5f5;
 --scroll-fg: #b3b3b3;
}
:root[theme="dark"], :root [data-jp-theme-light="false"]{
 --scroll-bg: #3c3c3c;
 --scroll-fg: #97e1fb;
}
body {
 scrollbar-color: var(--scroll-fg) var(--scroll-bg);
}
body::-webkit-scrollbar {
 width: 10px; /* Mostly for vertical scrollbars */
 height: 10px; /* Mostly for horizontal scrollbars */
}
body::-webkit-scrollbar-thumb {
 background-color: var(--scroll-fg);
}
body::-webkit-scrollbar-track {
 background-color: var(--scroll-bg);
}
 </style>
 </head>
 <body>
 
<table class="dataframe" id="df_-1426063308"></table>

<table class="dataframe" id="static_df_-1426063306"><thead><tr><th>most_starred_repo</th></tr></thead><tbody><tr><td>{ id:3432266, node_id:MDEwOlJlcG9zaXR<span class="structural">...</span></td></tr></tbody></table>
<p class="dataframe_description">DataFrame: rowsCount = 1, columnsCount = 37</p>

 </body>
 <script>
 (function () {
 window.DataFra

To make the result more obvious, let's create a new dataframe with a subset of columns from repos using `select`. Then we will use `head` to check the first 5 rows in the dataframe. 

In [22]:
organizations[0].repos.select { name and html_url and stargazers_count }.head(5)

<iframe onload="o_resize_iframe_out_10()" style="width:100%;" class="result_container" id="iframe_out_10" frameBorder="0" srcdoc=" <html theme='dark'>
 <head>
 <style type="text/css">
 :root {
 --background: #fff;
 --background-odd: #f5f5f5;
 --background-hover: #d9edfd;
 --header-text-color: #474747;
 --text-color: #848484;
 --text-color-dark: #000;
 --text-color-medium: #737373;
 --text-color-pale: #b3b3b3;
 --inner-border-color: #aaa;
 --bold-border-color: #000;
 --link-color: #296eaa;
 --link-color-pale: #296eaa;
 --link-hover: #1a466c;
}

:root[theme="dark"], :root [data-jp-theme-light="false"], .dataframe_dark{
 --background: #303030;
 --background-odd: #3c3c3c;
 --background-hover: #464646;
 --header-text-color: #dddddd;
 --text-color: #b3b3b3;
 --text-color-dark: #dddddd;
 --text-color-medium: #b2b2b2;
 --text-color-pale: #737373;
 --inner-border-color: #707070;
 --bold-border-color: #777777;
 --link-color: #008dc0;
 --link-color-pale: #97e1fb;
 --link-hover: #00688e;
}

p.dataframe_description {
 color: var(--text-color-dark);
}

table.dataframe {
 font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
 font-size: 12px;
 background-color: var(--background);
 color: var(--text-color-dark);
 border: none;
 border-collapse: collapse;
}

table.dataframe th, td {
 padding: 6px;
 border: 1px solid transparent;
 text-align: left;
}

table.dataframe th {
 background-color: var(--background);
 color: var(--header-text-color);
}

table.dataframe td {
 vertical-align: top;
}

table.dataframe th.bottomBorder {
 border-bottom-color: var(--bold-border-color);
}

table.dataframe tbody > tr:nth-child(odd) {
 background: var(--background-odd);
}

table.dataframe tbody > tr:nth-child(even) {
 background: var(--background);
}

table.dataframe tbody > tr:hover {
 background: var(--background-hover);
}

table.dataframe a {
 cursor: pointer;
 color: var(--link-color);
 text-decoration: none;
}

table.dataframe tr:hover > td a {
 color: var(--link-color-pale);
}

table.dataframe a:hover {
 color: var(--link-hover);
 text-decoration: underline;
}

table.dataframe img {
 max-width: fit-content;
}

table.dataframe th.complex {
 background-color: var(--background);
 border: 1px solid var(--background);
}

table.dataframe .leftBorder {
 border-left-color: var(--inner-border-color);
}

table.dataframe .rightBorder {
 border-right-color: var(--inner-border-color);
}

table.dataframe .rightAlign {
 text-align: right;
}

table.dataframe .expanderSvg {
 width: 8px;
 height: 8px;
 margin-right: 3px;
}

table.dataframe .expander {
 display: flex;
 align-items: center;
}

/* formatting */

table.dataframe .null {
 color: var(--text-color-pale);
}

table.dataframe .structural {
 color: var(--text-color-medium);
 font-weight: bold;
}

table.dataframe .dataFrameCaption {
 font-weight: bold;
}

table.dataframe .numbers {
 color: var(--text-color-dark);
}

table.dataframe td:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}

table.dataframe tr:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}





:root {
 --scroll-bg: #f5f5f5;
 --scroll-fg: #b3b3b3;
}
:root[theme="dark"], :root [data-jp-theme-light="false"]{
 --scroll-bg: #3c3c3c;
 --scroll-fg: #97e1fb;
}
body {
 scrollbar-color: var(--scroll-fg) var(--scroll-bg);
}
body::-webkit-scrollbar {
 width: 10px; /* Mostly for vertical scrollbars */
 height: 10px; /* Mostly for horizontal scrollbars */
}
body::-webkit-scrollbar-thumb {
 background-color: var(--scroll-fg);
}
body::-webkit-scrollbar-track {
 background-color: var(--scroll-bg);
}
 </style>
 </head>
 <body>
 
<table class="dataframe" id="df_-1426063305"></table>

<table class="dataframe" id="static_df_-1426063304"><thead><tr><th>name</th><th>html_url</th><th>stargazers_count</th></tr></thead><tbody><tr><td>JPS</td><td>https://github.com/JetBrains/JPS</td><td>23</td></tr><tr><td>YouTrackSharp</td><td>https://github.com/JetBrains/YouTrack<span class="structural">...</span></td><td>115</td></tr><tr><td>col

Looks better, doesn't it? Let's combine two previous examples

In [23]:
organizations.mapToColumn("most_starred_repo") {
    repos.select { name and html_url and stargazers_count }.maxBy { stargazers_count }
}

<iframe onload="o_resize_iframe_out_11()" style="width:100%;" class="result_container" id="iframe_out_11" frameBorder="0" srcdoc=" <html theme='dark'>
 <head>
 <style type="text/css">
 :root {
 --background: #fff;
 --background-odd: #f5f5f5;
 --background-hover: #d9edfd;
 --header-text-color: #474747;
 --text-color: #848484;
 --text-color-dark: #000;
 --text-color-medium: #737373;
 --text-color-pale: #b3b3b3;
 --inner-border-color: #aaa;
 --bold-border-color: #000;
 --link-color: #296eaa;
 --link-color-pale: #296eaa;
 --link-hover: #1a466c;
}

:root[theme="dark"], :root [data-jp-theme-light="false"], .dataframe_dark{
 --background: #303030;
 --background-odd: #3c3c3c;
 --background-hover: #464646;
 --header-text-color: #dddddd;
 --text-color: #b3b3b3;
 --text-color-dark: #dddddd;
 --text-color-medium: #b2b2b2;
 --text-color-pale: #737373;
 --inner-border-color: #707070;
 --bold-border-color: #777777;
 --link-color: #008dc0;
 --link-color-pale: #97e1fb;
 --link-hover: #00688e;
}

p.dataframe_description {
 color: var(--text-color-dark);
}

table.dataframe {
 font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
 font-size: 12px;
 background-color: var(--background);
 color: var(--text-color-dark);
 border: none;
 border-collapse: collapse;
}

table.dataframe th, td {
 padding: 6px;
 border: 1px solid transparent;
 text-align: left;
}

table.dataframe th {
 background-color: var(--background);
 color: var(--header-text-color);
}

table.dataframe td {
 vertical-align: top;
}

table.dataframe th.bottomBorder {
 border-bottom-color: var(--bold-border-color);
}

table.dataframe tbody > tr:nth-child(odd) {
 background: var(--background-odd);
}

table.dataframe tbody > tr:nth-child(even) {
 background: var(--background);
}

table.dataframe tbody > tr:hover {
 background: var(--background-hover);
}

table.dataframe a {
 cursor: pointer;
 color: var(--link-color);
 text-decoration: none;
}

table.dataframe tr:hover > td a {
 color: var(--link-color-pale);
}

table.dataframe a:hover {
 color: var(--link-hover);
 text-decoration: underline;
}

table.dataframe img {
 max-width: fit-content;
}

table.dataframe th.complex {
 background-color: var(--background);
 border: 1px solid var(--background);
}

table.dataframe .leftBorder {
 border-left-color: var(--inner-border-color);
}

table.dataframe .rightBorder {
 border-right-color: var(--inner-border-color);
}

table.dataframe .rightAlign {
 text-align: right;
}

table.dataframe .expanderSvg {
 width: 8px;
 height: 8px;
 margin-right: 3px;
}

table.dataframe .expander {
 display: flex;
 align-items: center;
}

/* formatting */

table.dataframe .null {
 color: var(--text-color-pale);
}

table.dataframe .structural {
 color: var(--text-color-medium);
 font-weight: bold;
}

table.dataframe .dataFrameCaption {
 font-weight: bold;
}

table.dataframe .numbers {
 color: var(--text-color-dark);
}

table.dataframe td:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}

table.dataframe tr:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}





:root {
 --scroll-bg: #f5f5f5;
 --scroll-fg: #b3b3b3;
}
:root[theme="dark"], :root [data-jp-theme-light="false"]{
 --scroll-bg: #3c3c3c;
 --scroll-fg: #97e1fb;
}
body {
 scrollbar-color: var(--scroll-fg) var(--scroll-bg);
}
body::-webkit-scrollbar {
 width: 10px; /* Mostly for vertical scrollbars */
 height: 10px; /* Mostly for horizontal scrollbars */
}
body::-webkit-scrollbar-thumb {
 background-color: var(--scroll-fg);
}
body::-webkit-scrollbar-track {
 background-color: var(--scroll-bg);
}
 </style>
 </head>
 <body>
 
<table class="dataframe" id="df_-1426063303"></table>

<table class="dataframe" id="static_df_-1426063302"><thead><tr><th>most_starred_repo</th></tr></thead><tbody><tr><td>{ name:kotlin, html_url:https://githu<span class="structural">...</span></td></tr></tbody></table>
<p class="dataframe_description">DataFrame: rowsCount = 1, columnsCount = 3</p>

 </body>
 <script>
 (function () {
 window.DataFr

3a) What is **the most and the least** starred project? There is a special syntax for `map` when you need to compute more than one column. This `map` creates a new dataframe.

In [24]:
organizations.mapToFrame { 
    "most_starred_repo" from { 
        repos
            .select { name and html_url and stargazers_count }
            .maxBy { stargazers_count } 
    }
    "least_starred_repo" from { 
        repos
            .select { name and html_url and stargazers_count }
            .minBy { stargazers_count } 
    }
}

<iframe onload="o_resize_iframe_out_12()" style="width:100%;" class="result_container" id="iframe_out_12" frameBorder="0" srcdoc=" <html theme='dark'>
 <head>
 <style type="text/css">
 :root {
 --background: #fff;
 --background-odd: #f5f5f5;
 --background-hover: #d9edfd;
 --header-text-color: #474747;
 --text-color: #848484;
 --text-color-dark: #000;
 --text-color-medium: #737373;
 --text-color-pale: #b3b3b3;
 --inner-border-color: #aaa;
 --bold-border-color: #000;
 --link-color: #296eaa;
 --link-color-pale: #296eaa;
 --link-hover: #1a466c;
}

:root[theme="dark"], :root [data-jp-theme-light="false"], .dataframe_dark{
 --background: #303030;
 --background-odd: #3c3c3c;
 --background-hover: #464646;
 --header-text-color: #dddddd;
 --text-color: #b3b3b3;
 --text-color-dark: #dddddd;
 --text-color-medium: #b2b2b2;
 --text-color-pale: #737373;
 --inner-border-color: #707070;
 --bold-border-color: #777777;
 --link-color: #008dc0;
 --link-color-pale: #97e1fb;
 --link-hover: #00688e;
}

p.dataframe_description {
 color: var(--text-color-dark);
}

table.dataframe {
 font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
 font-size: 12px;
 background-color: var(--background);
 color: var(--text-color-dark);
 border: none;
 border-collapse: collapse;
}

table.dataframe th, td {
 padding: 6px;
 border: 1px solid transparent;
 text-align: left;
}

table.dataframe th {
 background-color: var(--background);
 color: var(--header-text-color);
}

table.dataframe td {
 vertical-align: top;
}

table.dataframe th.bottomBorder {
 border-bottom-color: var(--bold-border-color);
}

table.dataframe tbody > tr:nth-child(odd) {
 background: var(--background-odd);
}

table.dataframe tbody > tr:nth-child(even) {
 background: var(--background);
}

table.dataframe tbody > tr:hover {
 background: var(--background-hover);
}

table.dataframe a {
 cursor: pointer;
 color: var(--link-color);
 text-decoration: none;
}

table.dataframe tr:hover > td a {
 color: var(--link-color-pale);
}

table.dataframe a:hover {
 color: var(--link-hover);
 text-decoration: underline;
}

table.dataframe img {
 max-width: fit-content;
}

table.dataframe th.complex {
 background-color: var(--background);
 border: 1px solid var(--background);
}

table.dataframe .leftBorder {
 border-left-color: var(--inner-border-color);
}

table.dataframe .rightBorder {
 border-right-color: var(--inner-border-color);
}

table.dataframe .rightAlign {
 text-align: right;
}

table.dataframe .expanderSvg {
 width: 8px;
 height: 8px;
 margin-right: 3px;
}

table.dataframe .expander {
 display: flex;
 align-items: center;
}

/* formatting */

table.dataframe .null {
 color: var(--text-color-pale);
}

table.dataframe .structural {
 color: var(--text-color-medium);
 font-weight: bold;
}

table.dataframe .dataFrameCaption {
 font-weight: bold;
}

table.dataframe .numbers {
 color: var(--text-color-dark);
}

table.dataframe td:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}

table.dataframe tr:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}





:root {
 --scroll-bg: #f5f5f5;
 --scroll-fg: #b3b3b3;
}
:root[theme="dark"], :root [data-jp-theme-light="false"]{
 --scroll-bg: #3c3c3c;
 --scroll-fg: #97e1fb;
}
body {
 scrollbar-color: var(--scroll-fg) var(--scroll-bg);
}
body::-webkit-scrollbar {
 width: 10px; /* Mostly for vertical scrollbars */
 height: 10px; /* Mostly for horizontal scrollbars */
}
body::-webkit-scrollbar-thumb {
 background-color: var(--scroll-fg);
}
body::-webkit-scrollbar-track {
 background-color: var(--scroll-bg);
}
 </style>
 </head>
 <body>
 
<table class="dataframe" id="df_-1426063301"></table>

<table class="dataframe" id="static_df_-1426063300"><thead><tr><th>most_starred_repo</th><th>least_starred_repo</th></tr></thead><tbody><tr><td>{ name:kotlin, html_url:https://githu<span class="structural">...</span></td><td>{ name:hsql1, html_url:https://github<span class="structural">...</span></td></tr></tbody></table>
<p class="dataframe_de

We can observe that we have to write the same select twice. Instead of this, let's **create a new dataframe, where column `repos`  has only 3 child columns: `name and html_url and stargazers_count`** and then get the most and least starred project again.

Here we want to modify values inside the dataframe, and the documentation suggests we can use [`update`](https://kotlin.github.io/dataframe/update.html) for this task.

In [25]:
val organizationsWithMinifiedRepos = organizations.update {
    // We want to change repos, so pick this column
    repos 
}
.with { 
    // Here *it* is *repos*, and we can use `select` as we did in the previous example.
    // Value of the `repos` column will be replaced by the result of the following expression
    it.select { name and html_url and stargazers_count } 
}   

Note that **dataframe is an immutable structure** and operators do not change values inside original dataframe. 

In [26]:
println(organizations.repos[0].columnNames())
println()
println(organizationsWithMinifiedRepos.repos[0].columnNames())

[id, node_id, name, full_name, private, html_url, description, fork, url, created_at, updated_at, pushed_at, homepage, size, stargazers_count, watchers_count, language, has_issues, has_projects, has_downloads, has_wiki, has_pages, forks_count, archived, disabled, open_issues_count, allow_forking, is_template, topics, visibility, forks, open_issues, watchers, default_branch, permissions, contributors, license]

[name, html_url, stargazers_count]


Here is the answer for **what are the most and the least starred projects**.

In [27]:
organizationsWithMinifiedRepos.mapToFrame { 
    "most_starred_repo" from { 
        repos.maxBy { stargazers_count } 
    }
    "least_starred_repo" from { 
        repos.minBy { stargazers_count } 
    }
}

<iframe onload="o_resize_iframe_out_13()" style="width:100%;" class="result_container" id="iframe_out_13" frameBorder="0" srcdoc=" <html theme='dark'>
 <head>
 <style type="text/css">
 :root {
 --background: #fff;
 --background-odd: #f5f5f5;
 --background-hover: #d9edfd;
 --header-text-color: #474747;
 --text-color: #848484;
 --text-color-dark: #000;
 --text-color-medium: #737373;
 --text-color-pale: #b3b3b3;
 --inner-border-color: #aaa;
 --bold-border-color: #000;
 --link-color: #296eaa;
 --link-color-pale: #296eaa;
 --link-hover: #1a466c;
}

:root[theme="dark"], :root [data-jp-theme-light="false"], .dataframe_dark{
 --background: #303030;
 --background-odd: #3c3c3c;
 --background-hover: #464646;
 --header-text-color: #dddddd;
 --text-color: #b3b3b3;
 --text-color-dark: #dddddd;
 --text-color-medium: #b2b2b2;
 --text-color-pale: #737373;
 --inner-border-color: #707070;
 --bold-border-color: #777777;
 --link-color: #008dc0;
 --link-color-pale: #97e1fb;
 --link-hover: #00688e;
}

p.dataframe_description {
 color: var(--text-color-dark);
}

table.dataframe {
 font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
 font-size: 12px;
 background-color: var(--background);
 color: var(--text-color-dark);
 border: none;
 border-collapse: collapse;
}

table.dataframe th, td {
 padding: 6px;
 border: 1px solid transparent;
 text-align: left;
}

table.dataframe th {
 background-color: var(--background);
 color: var(--header-text-color);
}

table.dataframe td {
 vertical-align: top;
}

table.dataframe th.bottomBorder {
 border-bottom-color: var(--bold-border-color);
}

table.dataframe tbody > tr:nth-child(odd) {
 background: var(--background-odd);
}

table.dataframe tbody > tr:nth-child(even) {
 background: var(--background);
}

table.dataframe tbody > tr:hover {
 background: var(--background-hover);
}

table.dataframe a {
 cursor: pointer;
 color: var(--link-color);
 text-decoration: none;
}

table.dataframe tr:hover > td a {
 color: var(--link-color-pale);
}

table.dataframe a:hover {
 color: var(--link-hover);
 text-decoration: underline;
}

table.dataframe img {
 max-width: fit-content;
}

table.dataframe th.complex {
 background-color: var(--background);
 border: 1px solid var(--background);
}

table.dataframe .leftBorder {
 border-left-color: var(--inner-border-color);
}

table.dataframe .rightBorder {
 border-right-color: var(--inner-border-color);
}

table.dataframe .rightAlign {
 text-align: right;
}

table.dataframe .expanderSvg {
 width: 8px;
 height: 8px;
 margin-right: 3px;
}

table.dataframe .expander {
 display: flex;
 align-items: center;
}

/* formatting */

table.dataframe .null {
 color: var(--text-color-pale);
}

table.dataframe .structural {
 color: var(--text-color-medium);
 font-weight: bold;
}

table.dataframe .dataFrameCaption {
 font-weight: bold;
}

table.dataframe .numbers {
 color: var(--text-color-dark);
}

table.dataframe td:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}

table.dataframe tr:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}





:root {
 --scroll-bg: #f5f5f5;
 --scroll-fg: #b3b3b3;
}
:root[theme="dark"], :root [data-jp-theme-light="false"]{
 --scroll-bg: #3c3c3c;
 --scroll-fg: #97e1fb;
}
body {
 scrollbar-color: var(--scroll-fg) var(--scroll-bg);
}
body::-webkit-scrollbar {
 width: 10px; /* Mostly for vertical scrollbars */
 height: 10px; /* Mostly for horizontal scrollbars */
}
body::-webkit-scrollbar-thumb {
 background-color: var(--scroll-fg);
}
body::-webkit-scrollbar-track {
 background-color: var(--scroll-bg);
}
 </style>
 </head>
 <body>
 
<table class="dataframe" id="df_-1426063299"></table>

<table class="dataframe" id="static_df_-1426063298"><thead><tr><th>most_starred_repo</th><th>least_starred_repo</th></tr></thead><tbody><tr><td>{ name:kotlin, html_url:https://githu<span class="structural">...</span></td><td>{ name:hsql1, html_url:https://github<span class="structural">...</span></td></tr></tbody></table>
<p class="dataframe_de

Expressions inside map, the [**row expressions**](https://kotlin.github.io/dataframe/datarow.html#row-expressions), can be as complex as your fantasy goes. For example, we can pick the five most starred and five least starred repositories.

In [28]:
organizationsWithMinifiedRepos.mapToColumn("interesting_repos") {
    val sorted = repos.sortByDesc { stargazers_count }
    sorted.take(5).concat(sorted.takeLast(5))
}

<iframe onload="o_resize_iframe_out_14()" style="width:100%;" class="result_container" id="iframe_out_14" frameBorder="0" srcdoc=" <html theme='dark'>
 <head>
 <style type="text/css">
 :root {
 --background: #fff;
 --background-odd: #f5f5f5;
 --background-hover: #d9edfd;
 --header-text-color: #474747;
 --text-color: #848484;
 --text-color-dark: #000;
 --text-color-medium: #737373;
 --text-color-pale: #b3b3b3;
 --inner-border-color: #aaa;
 --bold-border-color: #000;
 --link-color: #296eaa;
 --link-color-pale: #296eaa;
 --link-hover: #1a466c;
}

:root[theme="dark"], :root [data-jp-theme-light="false"], .dataframe_dark{
 --background: #303030;
 --background-odd: #3c3c3c;
 --background-hover: #464646;
 --header-text-color: #dddddd;
 --text-color: #b3b3b3;
 --text-color-dark: #dddddd;
 --text-color-medium: #b2b2b2;
 --text-color-pale: #737373;
 --inner-border-color: #707070;
 --bold-border-color: #777777;
 --link-color: #008dc0;
 --link-color-pale: #97e1fb;
 --link-hover: #00688e;
}

p.dataframe_description {
 color: var(--text-color-dark);
}

table.dataframe {
 font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
 font-size: 12px;
 background-color: var(--background);
 color: var(--text-color-dark);
 border: none;
 border-collapse: collapse;
}

table.dataframe th, td {
 padding: 6px;
 border: 1px solid transparent;
 text-align: left;
}

table.dataframe th {
 background-color: var(--background);
 color: var(--header-text-color);
}

table.dataframe td {
 vertical-align: top;
}

table.dataframe th.bottomBorder {
 border-bottom-color: var(--bold-border-color);
}

table.dataframe tbody > tr:nth-child(odd) {
 background: var(--background-odd);
}

table.dataframe tbody > tr:nth-child(even) {
 background: var(--background);
}

table.dataframe tbody > tr:hover {
 background: var(--background-hover);
}

table.dataframe a {
 cursor: pointer;
 color: var(--link-color);
 text-decoration: none;
}

table.dataframe tr:hover > td a {
 color: var(--link-color-pale);
}

table.dataframe a:hover {
 color: var(--link-hover);
 text-decoration: underline;
}

table.dataframe img {
 max-width: fit-content;
}

table.dataframe th.complex {
 background-color: var(--background);
 border: 1px solid var(--background);
}

table.dataframe .leftBorder {
 border-left-color: var(--inner-border-color);
}

table.dataframe .rightBorder {
 border-right-color: var(--inner-border-color);
}

table.dataframe .rightAlign {
 text-align: right;
}

table.dataframe .expanderSvg {
 width: 8px;
 height: 8px;
 margin-right: 3px;
}

table.dataframe .expander {
 display: flex;
 align-items: center;
}

/* formatting */

table.dataframe .null {
 color: var(--text-color-pale);
}

table.dataframe .structural {
 color: var(--text-color-medium);
 font-weight: bold;
}

table.dataframe .dataFrameCaption {
 font-weight: bold;
}

table.dataframe .numbers {
 color: var(--text-color-dark);
}

table.dataframe td:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}

table.dataframe tr:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}





:root {
 --scroll-bg: #f5f5f5;
 --scroll-fg: #b3b3b3;
}
:root[theme="dark"], :root [data-jp-theme-light="false"]{
 --scroll-bg: #3c3c3c;
 --scroll-fg: #97e1fb;
}
body {
 scrollbar-color: var(--scroll-fg) var(--scroll-bg);
}
body::-webkit-scrollbar {
 width: 10px; /* Mostly for vertical scrollbars */
 height: 10px; /* Mostly for horizontal scrollbars */
}
body::-webkit-scrollbar-thumb {
 background-color: var(--scroll-fg);
}
body::-webkit-scrollbar-track {
 background-color: var(--scroll-bg);
}
 </style>
 </head>
 <body>
 
<table class="dataframe" id="df_-1426063297"></table>

<table class="dataframe" id="static_df_-1426063295"><thead><tr><th>interesting_repos</th></tr></thead><tbody><tr><td>[10 x 3]</td></tr></tbody></table>
<p class="dataframe_description">DataColumn: name = "interesting_repos", type = DataFrame<*>, size = 1</p>

 </body>
 <script>
 (function () {
 window.DataFrame = window.DataFrame || new (functi

That's it. We used the following operations to explore our data sample:
* [`map`](https://kotlin.github.io/dataframe/map.html) to create a data column and dataframe with values computed from rows of original dataframe
* [`select`](https://kotlin.github.io/dataframe/columnselectors.html) to create dataframe with subset of columns from original
* [`convert`](https://kotlin.github.io/dataframe/convert.html) to create dataframe with changed values in some columns (in our case it was `repos`)

Here is some code required to work with GitHub API with API key. Without API the key, there is a limit of 30 requests or so. 
Follow the instruction to get a personal access token: https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token. Then update line `val credentials: Credentials? = null` like `val credentials: Credentials? = Credentials("username", "your_token")`

In [29]:
@file:DependsOn("org.http4k:http4k-core:4.16.3.0")
@file:DependsOn("org.http4k:http4k-client-apache:4.16.3.0")

In [30]:
import org.http4k.client.ApacheClient
import org.http4k.core.Method.GET
import org.http4k.core.Request

val client = ApacheClient()

In [31]:
data class Credentials(val name: String, val token: String)

In [32]:
import java.util.Base64

fun authHeader(name: String, token: String): String {
    return "Basic " + Base64.getEncoder().encode("$name:$token".toByteArray()).toString(Charsets.UTF_8)
}

fun readDf(url: String, credentials: Credentials?): DataFrame<*> {
    val request = Request(GET, url).let {
        if (credentials != null) {
            it.header("Authorization", authHeader(credentials.name, credentials.token))
        } else {
            it
        }
    }.query("per_page", "100")
    return DataFrame.readJsonStr(client(request).bodyString())
}

In [33]:
val credentials: Credentials? = null

fun readDf(url: String) = readDf(url, credentials)

In [34]:
import org.http4k.core.HttpHandler

data class PageInfo(val body: DataFrame<*>, val pageUrls: List<String>)

fun parseLinkHeader(links: Pair<String, String?>): String? = links.second?.split(",")
        ?.map { it.split(";")
        .map { it.trim() } }
        ?.find { (_, rel) -> rel == "rel=\"last\"" }
        ?.first()
        ?.drop(1)?.dropLast(1)
        

val PAGE_REGEX = "page=(\\d+)$".toRegex()
fun readPageInfo(client: HttpHandler, url: String): PageInfo {
    val firstPage = client(Request(GET, url).query("per_page", "100"))
    val firstPageDf = DataFrame.readJsonStr(firstPage.bodyString())
    val links = firstPage.headers.find { (name, _) -> name == "Link" } ?: return PageInfo(firstPageDf, emptyList())
    val lastPageUrl = parseLinkHeader(links) ?: error("Expected 'Link' header with non empty value 'rel=last' attribute on endpoint $url, but was $links")
    val n = PAGE_REGEX.find(lastPageUrl)?.groupValues?.get(1)?.toInt()!!
    val pageUrls = (2..n).map {
        lastPageUrl.replace(PAGE_REGEX, "page=$it")    
    }
    return PageInfo(firstPageDf, pageUrls)
}

In [35]:
fun readPaginatedEndpoint(url: String): DataFrame<*> {
    val pageInfo = readPageInfo(client, url)
    return pageInfo.body.concat(pageInfo.pageUrls.map { readDf(it) }.concat())
}

Section 2. Build the data sample

In [36]:
val jb = readDf("https://api.github.com/orgs/JetBrains")
jb

<iframe onload="o_resize_iframe_out_15()" style="width:100%;" class="result_container" id="iframe_out_15" frameBorder="0" srcdoc=" <html theme='dark'>
 <head>
 <style type="text/css">
 :root {
 --background: #fff;
 --background-odd: #f5f5f5;
 --background-hover: #d9edfd;
 --header-text-color: #474747;
 --text-color: #848484;
 --text-color-dark: #000;
 --text-color-medium: #737373;
 --text-color-pale: #b3b3b3;
 --inner-border-color: #aaa;
 --bold-border-color: #000;
 --link-color: #296eaa;
 --link-color-pale: #296eaa;
 --link-hover: #1a466c;
}

:root[theme="dark"], :root [data-jp-theme-light="false"], .dataframe_dark{
 --background: #303030;
 --background-odd: #3c3c3c;
 --background-hover: #464646;
 --header-text-color: #dddddd;
 --text-color: #b3b3b3;
 --text-color-dark: #dddddd;
 --text-color-medium: #b2b2b2;
 --text-color-pale: #737373;
 --inner-border-color: #707070;
 --bold-border-color: #777777;
 --link-color: #008dc0;
 --link-color-pale: #97e1fb;
 --link-hover: #00688e;
}

p.dataframe_description {
 color: var(--text-color-dark);
}

table.dataframe {
 font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
 font-size: 12px;
 background-color: var(--background);
 color: var(--text-color-dark);
 border: none;
 border-collapse: collapse;
}

table.dataframe th, td {
 padding: 6px;
 border: 1px solid transparent;
 text-align: left;
}

table.dataframe th {
 background-color: var(--background);
 color: var(--header-text-color);
}

table.dataframe td {
 vertical-align: top;
}

table.dataframe th.bottomBorder {
 border-bottom-color: var(--bold-border-color);
}

table.dataframe tbody > tr:nth-child(odd) {
 background: var(--background-odd);
}

table.dataframe tbody > tr:nth-child(even) {
 background: var(--background);
}

table.dataframe tbody > tr:hover {
 background: var(--background-hover);
}

table.dataframe a {
 cursor: pointer;
 color: var(--link-color);
 text-decoration: none;
}

table.dataframe tr:hover > td a {
 color: var(--link-color-pale);
}

table.dataframe a:hover {
 color: var(--link-hover);
 text-decoration: underline;
}

table.dataframe img {
 max-width: fit-content;
}

table.dataframe th.complex {
 background-color: var(--background);
 border: 1px solid var(--background);
}

table.dataframe .leftBorder {
 border-left-color: var(--inner-border-color);
}

table.dataframe .rightBorder {
 border-right-color: var(--inner-border-color);
}

table.dataframe .rightAlign {
 text-align: right;
}

table.dataframe .expanderSvg {
 width: 8px;
 height: 8px;
 margin-right: 3px;
}

table.dataframe .expander {
 display: flex;
 align-items: center;
}

/* formatting */

table.dataframe .null {
 color: var(--text-color-pale);
}

table.dataframe .structural {
 color: var(--text-color-medium);
 font-weight: bold;
}

table.dataframe .dataFrameCaption {
 font-weight: bold;
}

table.dataframe .numbers {
 color: var(--text-color-dark);
}

table.dataframe td:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}

table.dataframe tr:hover .formatted .structural, .null {
 color: var(--text-color-dark);
}





:root {
 --scroll-bg: #f5f5f5;
 --scroll-fg: #b3b3b3;
}
:root[theme="dark"], :root [data-jp-theme-light="false"]{
 --scroll-bg: #3c3c3c;
 --scroll-fg: #97e1fb;
}
body {
 scrollbar-color: var(--scroll-fg) var(--scroll-bg);
}
body::-webkit-scrollbar {
 width: 10px; /* Mostly for vertical scrollbars */
 height: 10px; /* Mostly for horizontal scrollbars */
}
body::-webkit-scrollbar-thumb {
 background-color: var(--scroll-fg);
}
body::-webkit-scrollbar-track {
 background-color: var(--scroll-bg);
}
 </style>
 </head>
 <body>
 
<table class="dataframe" id="df_-1426063294"></table>

<table class="dataframe" id="static_df_-1426063293"><thead><tr><th>message</th><th>documentation_url</th></tr></thead><tbody><tr><td>API rate limit exceeded for 83.84.95.<span class="structural">...</span></td><td>https://docs.github.com/rest/overview<span class="structural">...</span></td></tr></tbody></table>
<p class="dataframe_description">

In [37]:
jb.schema()

message: String
documentation_url: String

Most of the data in the dataframe are just URLs. 

In [38]:
val jb1 = jb
    .add("repos") { readDf(repos_url) }
    .add("members") { readPaginatedEndpoint("${url}/public_members") }
jb1

Line_62.jupyter.kts (2:28 - 37) Unresolved reference: repos_url
Line_62.jupyter.kts (3:48 - 51) Unresolved reference. None of the following candidates is applicable because of receiver type mismatch: 
public final val ColumnsContainer<Line_46_jupyter._DataFrameType17>.url: DataColumn<URL> defined in Line_46_jupyter
public final val ColumnsContainer<Line_46_jupyter._DataFrameType17?>.url: DataColumn<URL?> defined in Line_46_jupyter
public final val DataRow<Line_46_jupyter._DataFrameType17>.url: URL defined in Line_46_jupyter
public final val DataRow<Line_46_jupyter._DataFrameType17?>.url: URL? defined in Line_46_jupyter

But now we have repositories, each repository has contributors_url. If we want to download contributors, we need to add a new column to `repos`. 

In [39]:
val jb2 = jb1
    .convert { repos }.with { 
        it.add("contributors") { readDf(contributors_url) }
    }

Line_63.jupyter.kts (1:11 - 14) Unresolved reference: jb1
Line_63.jupyter.kts (2:16 - 21) Unresolved reference. None of the following candidates is applicable because of receiver type mismatch: 
public final val ColumnsContainer<Line_46_jupyter._DataFrameType17>.repos: DataColumn<DataFrame<Line_46_jupyter._DataFrameType18>> defined in Line_46_jupyter
public final val ColumnsContainer<Line_46_jupyter._DataFrameType17?>.repos: DataColumn<DataFrame<Line_46_jupyter._DataFrameType18?>> defined in Line_46_jupyter
public final val DataRow<Line_46_jupyter._DataFrameType17>.repos: DataFrame<Line_46_jupyter._DataFrameType18> defined in Line_46_jupyter
public final val DataRow<Line_46_jupyter._DataFrameType17?>.repos: DataFrame<Line_46_jupyter._DataFrameType18?> defined in Line_46_jupyter
Line_63.jupyter.kts (3:12 - 15) Unresolved reference: add
Line_63.jupyter.kts (3:41 - 57) Unresolved reference: contributors_url

Unfortunately, it gets worse when you need to add a new child column to some deeply nested column. Like, for example, downloading a list of followers for each contributor of each repository would require writing this code:
```
jb2.convert { repos }.with {
    it.convert { 
        contributors 
    }.with { 
        it.add("followers") { readDf(followers_url) } 
    }
}
```
You can still do it, if you need to.

In [40]:
jb2.schema()

Line_64.jupyter.kts (1:1 - 4) Unresolved reference: jb2

Resulting dataframe has a lot of columns. We can use `select` and `remove` to filter them. Both operations provide DSL for selecting arbitrary set of columns https://kotlin.github.io/dataframe/columnselectors.html

In [41]:
val minifiedSample = jb2
    .select { cols(url, name, repos, members) }
    // Like before, we use convert to create dataframe with changed values in nested column
    .convert { members }.with { it.remove { endsWith("_url") } }
    .convert { repos }.with { it.remove { endsWith("_url").except(html_url) } }
    .convert { repos }.with { it.remove { owner } }
    .convert { repos }.with { 
        it.convert { contributors }.with { it.remove { endsWith("_url") } } 
    }

Line_65.jupyter.kts (1:22 - 25) Unresolved reference: jb2
Line_65.jupyter.kts (2:15 - 19) Unresolved reference: cols
Line_65.jupyter.kts (2:20 - 23) Unresolved reference. None of the following candidates is applicable because of receiver type mismatch: 
public final val ColumnsContainer<Line_46_jupyter._DataFrameType17>.url: DataColumn<URL> defined in Line_46_jupyter
public final val ColumnsContainer<Line_46_jupyter._DataFrameType17?>.url: DataColumn<URL?> defined in Line_46_jupyter
public final val DataRow<Line_46_jupyter._DataFrameType17>.url: URL defined in Line_46_jupyter
public final val DataRow<Line_46_jupyter._DataFrameType17?>.url: URL? defined in Line_46_jupyter
Line_65.jupyter.kts (2:25 - 29) Unresolved reference. None of the following candidates is applicable because of receiver type mismatch: 
public final val ColumnsContainer<Line_46_jupyter._DataFrameType17>.name: DataColumn<String> defined in Line_46_jupyter
public final val ColumnsContainer<Line_46_jupyter._DataFrameTyp

In [42]:
minifiedSample.schema()

Line_66.jupyter.kts (1:1 - 15) Unresolved reference: minifiedSample

In [43]:
minifiedSample.writeJson("jetbrains.json")

Line_67.jupyter.kts (1:1 - 15) Unresolved reference: minifiedSample